src/shared/machine-credential.c

```diff
@@ -0,0 +1,135 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "alloc-util.h"
+#include "creds-util.h"
+#include "escape.h"
+#include "extract-word.h"
+#include "fileio.h"
+#include "macro.h"
+#include "memory-util.h"
+#include "machine-credential.h"
+#include "path-util.h"
+#include "string-util-fundamental.h"
+
+static void machine_credential_free(MachineCredential *cred) {
+        assert(cred);
+
+        cred->id = mfree(cred->id);
+        cred->data = erase_and_free(cred->data);
+        cred->size = 0;
+}
+
+void machine_credential_free_all(MachineCredential *creds, size_t n) {
+        size_t i;
+
+        assert(creds || n == 0);
+
+        for (i = 0; i < n; i++)
+                machine_credential_free(creds + i);
+
+        free(creds);
+}
+
+int machine_credential_set(MachineCredential **arg_credentials, size_t *arg_n_credentials, const char *cred_string) {
+        _cleanup_free_ char *word = NULL, *data = NULL;
+        MachineCredential *creds = *arg_credentials;
+        ssize_t l;
+        size_t n_creds = *arg_n_credentials;
+        int r;
+        const char *p = cred_string;
+
+        r = extract_first_word(&p, &word, ":", EXTRACT_DONT_COALESCE_SEPARATORS);
+        if (r == -ENOMEM)
+                return r;
+        if (r < 0)
+                return log_debug_errno(r, "Failed to parse --set-credential= parameter: %m");
+        if (r == 0 || !p)
+                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Missing value for --set-credential=: %s", cred_string);
+
+        if (!credential_name_valid(word))
+                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "MachineCredential name is not valid: %s", word);
+
+        for (size_t i = 0; i < n_creds; i++)
+                if (streq(creds[i].id, word))
+                        return log_debug_errno(SYNTHETIC_ERRNO(EEXIST), "Duplicate credential '%s', refusing.", word);
+
+        l = cunescape(p, UNESCAPE_ACCEPT_NUL, &data);
+        if (l < 0)
+                return log_debug_errno(l, "Failed to unescape credential data: %s", p);
+
+        creds = GREEDY_REALLOC(creds, n_creds + 1);
+        if (!creds)
+                return -ENOMEM;
+
+        creds[n_creds++] = (MachineCredential) {
+                .id = TAKE_PTR(word),
+                .data = TAKE_PTR(data),
+                .size = l,
+        };
+
+        *arg_credentials = creds;
+        *arg_n_credentials = n_creds;
+
+        return 0;
+}
+
+int machine_credential_load(MachineCredential **arg_credentials, size_t *arg_n_credentials, const char *cred_path) {
+        ReadFullFileFlags flags = READ_FULL_FILE_SECURE;
+        _cleanup_(erase_and_freep) char *data = NULL;
+        _cleanup_free_ char *word = NULL, *j = NULL;
+        MachineCredential *creds = *arg_credentials;
```

Comment:
This is a shallow copy, which means you're still operating on the original instance of `MachineCredential*` after all, and thus is unnecessary.

Reply:
I use a second variable to handle the result of the reallocarray in the case of failure.

I have assigned it here to use the shorter name everywhere and so that I don't have to manually dereference everywhere.

This also allows me to only assign back to these pointers in the case that the reallocaray succeeded.
This is required because if I assigned it immediately it would NULL out a pointer in the case of failure but the overwritten pointer would still be valid as rellocarray is defined to leave memory untouched in the case of failure.

Reply:
> This also allows me to only assign back to these pointers in the case that the reallocaray succeeded.
This is required because if I assigned it immediately it would NULL out a pointer in the case of failure but the overwritten pointer would still be valid as rellocarray is defined to leave memory untouched in the case of failure.

There's no need to assign the pointer returned by `GREEDY_REALLOC` to `creds`. This is handled gracefully internally.